x-pack/plugins/observability_ai_assistant/server/service/knowledge_base_service/index.ts

```diff
@@ -291,64 +292,202 @@ export class KnowledgeBaseService {
+              {
+                term: {
+                  [modelField]: modelId,
+                },
+              },
+            ],
+          },
+        };
+      });
+    });
+
+    const response = await asCurrentUser.search<unknown>({
+      index: 'search-*',
+      query: {
+        bool: {
+          should: esQueries,
+        },
+      },
+      size: 20,
+      _source: {
+        exclude: ['_*', 'ml*'],
+      },
+    });
+
+    return response.hits.hits.map((hit) => ({
+      text: JSON.stringify(hit._source),
+      score: hit._score!,
+      is_correction: false,
+      id: hit._id,
+    }));
+  }
+
   recall = async ({
     user,
     queries,
     contexts,
     namespace,
+    asCurrentUser,
   }: {
     queries: string[];
     contexts?: string[];
     user: { name: string };
     namespace: string;
+    asCurrentUser: ElasticsearchClient;
   }): Promise<{
     entries: RecalledEntry[];
   }> => {
-    try {
-      const query = {
-        bool: {
-          should: queries.map((text) => ({
-            text_expansion: {
-              'ml.tokens': {
-                model_text: text,
-                model_id: ELSER_MODEL_ID,
-              },
-            } as unknown as QueryDslTextExpansionQuery,
-          })),
-          filter: [
-            ...getAccessQuery({
-              user,
-              namespace,
-            }),
-            ...getCategoryQuery({ contexts }),
-          ],
-        },
-      };
-
-      const response = await this.dependencies.esClient.search<
-        Pick<KnowledgeBaseEntry, 'text' | 'is_correction' | 'labels'>
-      >({
-        index: [this.dependencies.resources.aliases.kb],
-        query,
-        size: 20,
-        _source: {
-          includes: ['text', 'is_correction', 'labels'],
-        },
-      });
+    const modelId = ELSER_MODEL_ID;
+
+    const [documentsFromKb, documentsFromConnectors] = await Promise.all([
+      this.recallFromKnowledgeBase({
+        user,
+        queries,
+        contexts,
+        namespace,
+        modelId,
+      }).catch((error) => {
+        if (isAlreadyExistsError(error)) {
+          throwKnowledgeBaseNotReady(error.body);
+        }
+        throw error;
+      }),
+      this.recallFromConnectors({
+        asCurrentUser,
+        queries,
+        modelId,
+      }).catch((error) => {
+        this.dependencies.logger.debug('Error getting data from search indices');
+        this.dependencies.logger.debug(error);
+        return [];
+      }),
+    ]);
+
+    const sortedEntries = orderBy(
+      documentsFromKb.concat(documentsFromConnectors),
+      'score',
+      'desc'
+    ).slice(0, 20);
+
+    const MAX_TOKENS = 4000;
+
+    let tokenCount = 0;
+
+    const returnedEntries: RecalledEntry[] = [];
+
+    let entry: RecalledEntry | undefined;
+    while ((entry = sortedEntries.shift()) && tokenCount <= MAX_TOKENS) {
+      returnedEntries.push(entry);
+      tokenCount += encode(entry.text).length;
+    }
```

Comment:
Not really a suggestion, was just thinking of another way to express this.

```suggestion
    const returnedEntries: RecalledEntry[] = [];
    sortedEntries.every((entry) => {
      const tokens = encode(entry.text).length;
      
      if (tokenCount + tokens > MAX_TOKENS) {
        return false;
      }
      
      returnedEntries.push(entry);
      tokenCount += tokens;
    });
```

One could also imagine reaching the token limit and checking if any of the lower ranking documents would fit in the remaining space but that is perhaps a silly optimisation to perform.

Reply:
It's probably better (except for swapping out `.every` with `.forEach`. I'll change it.

Reply:
With forEach you won't have a way to exit early, but you could just skip adding the items if you've already reached the limit

Reply:
oh yeah, good point, but I'd rather `for...of` in that case.